WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -0,0 +1,191 @@
+		private CredentialPool VsizeCredentialPool { get; } = new();
+		private Round Round { get; }
+		public IArenaRequestHandler ArenaRequestHandler { get; }
+		private BitcoinSecret BitcoinSecret { get; }
+		public KeyManager Keymanager { get; }
+		private SecureRandom SecureRandom { get; }
+		private CancellationTokenSource DisposeCts { get; } = new();
+		private Coin[] Coins { get; set; }
+
+		public CoinJoinClient(
+			Round round,
+			IArenaRequestHandler arenaRequestHandler,
+			BitcoinSecret bitcoinSecret,
+			IEnumerable<Coin> coins,
+			KeyManager keymanager)
+		{
+			Round = round;
+			ArenaRequestHandler = arenaRequestHandler;
+			BitcoinSecret = bitcoinSecret;
+			Keymanager = keymanager;
+			SecureRandom = new SecureRandom();
+			Coins = coins.ToArray();
+		}
+
+		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+		{
+			try
+			{
+				// Register coins.
+				AliceClient[] aliceClients = await RegisterCoinsAsync().ConfigureAwait(false);
+
+				// Confirm coins.
+				await ConfirmConnectionsAsync(aliceClients, stoppingToken).ConfigureAwait(false);
+
+				Money[]? othersInputAmounts = null; // TODO: Get it from somewhere.
+
+				// Planning
+				var outputs = await PlanAmountsAsync(othersInputAmounts, stoppingToken).ConfigureAwait(false);
+
+				// Output registration.
+				await RegisterOutputsAsync(outputs).ConfigureAwait(false);
+
+				Transaction? unsignedCoinJoinTransaction = null; // TODO: Get it from somewhere.
+
+				// Send signature.
+				await SignTransactionAsync(aliceClients, unsignedCoinJoinTransaction).ConfigureAwait(false);
+			}
+			catch (Exception ex)
+			{
+			}
+		}
+
+		public async Task StartMixingCoinsAsync()
+		{
+			await StartAsync(DisposeCts.Token).ConfigureAwait(false);
+		}
+
+		private async Task<AliceClient[]> RegisterCoinsAsync()
+		{
+			var aliceArenaClient = new ArenaClient(
+				Round.AmountCredentialIssuerParameters,
+				Round.VsizeCredentialIssuerParameters,
+				AmountCredentialPool,
+				VsizeCredentialPool,
+				ArenaRequestHandler,
+				SecureRandom);
+
+			List<AliceClient> aliceClients = new();
+			try
+			{
+				foreach (var coin in Coins)
+				{
+					// Parallelize or Random delay?
+					aliceClients.Add(await AliceClient.CreateNewAsync(aliceArenaClient, coin, BitcoinSecret, Round.Id, Round.Hash, Round.FeeRate).ConfigureAwait(false));
+				}
+			}
+			catch (Exception)
+			{
+				foreach (var alice in aliceClients)
+				{
+					// Remove already registered Inputs.
+					try
+					{
+						await alice.RemoveInputAsync().ConfigureAwait(false);
+					}
+					catch (Exception)
+					{
+						// Log?
+					}
+				}
+				throw;
+			}
+
+			return aliceClients.ToArray();
+		}
+
+		private async Task ConfirmConnectionsAsync(AliceClient[] aliceClients, CancellationToken stoppingToken)
+		{
+			Task[] confirmTasks = aliceClients.Select(aliceClient => aliceClient.ConfirmConnectionAsync(TimeSpan.FromSeconds(5), stoppingToken)).ToArray();
```

Comment:
Here we could need some delays too because otherwise the coordinator "could" (i am not sure really) correlate all the fast sequence of confirmations as belonging to the same wallet. 

This is just one idea:
```c#
var confirmTasks = aliceClients
    .Select(x => xConfirmConnectionAsync(...).WithRandomDelay())
```

```c#
async Task<T> WithRandomDelayAsync<T>(this Task<T> theTask)
{
    await Task.Delay(random);
    return theTask;
}
```

Reply:
The coordinator could do that quite easily, although if the client queries status on multiple tor circuits and checks status responses for consistency it gets harder.

Assuming the minimum number of status requests and no delays, the coordinator can choose to one request at a time to respond indicating that connection confirmation has begun, while reporting input registration to other responses.

Note that this is not theoretical, because even when checking consistency, the client must tolerate some delays and cannot poll at very high resolution. Since the coordinator can blamelessly DoS its own rounds, generating as many blame rounds as it wants, small timing correlation signals can be amplified.

In zkSNACKs/WabiSabi#83 I described how this can be mitigated, by forcing an honest coordinator to commit to a total ordering of status responses (by serving requests following a round robin circular list, and ensuring that each response commits to the request and the previous responses), but I need to go into a bit more detail about how exactly the commitment scheme works and how likely clients are to detect a deviation based on the number of circuits they are polling over.

See also https://github.com/zkSNACKs/WalletWasabi/pull/5694#discussion_r622478177 for additional considerations apart from such timing based leaks.